File: Chapter16/SampleFrameAnimation/Java/SampleFrameAnimation/app/src/main/java/com/ai/android/frameanimation/FrameAnimationActivity.java
package com.ai.android.frameanimation;

import androidx.appcompat.app.AppCompatActivity;

import android.graphics.drawable.AnimationDrawable;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.ImageView;

public class FrameAnimationActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.frame_animations_layout);
        setupButton();
    }

    private void setupButton() {
        Button b = findViewById(R.id.startFAButtonId);
        b.setOnClickListener(this::parentButtonClicked);
    }

    private void parentButtonClicked(View v) {
        animate();
    }

    private void animate() {
        ImageView imgView = findViewById(R.id.imageView);
        imgView.setVisibility(ImageView.VISIBLE);
        imgView.setBackgroundResource(R.drawable.frame_animation);

        AnimationDrawable frameAnimation = (AnimationDrawable) imgView.getBackground();
        if (frameAnimation.isRunning()) {
            frameAnimation.stop();
        } else {
            frameAnimation.stop();
            frameAnimation.start();
        }
    }

}

File: Chapter16/SampleFrameAnimation/Kotlin/SampleFrameAnimation/app/src/main/java/com/ai/android/frameanimation/FrameAnimationActivity.kt
package com.ai.android.frameanimation

import android.graphics.drawable.AnimationDrawable
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.view.View
import android.widget.Button
import android.widget.ImageView

class FrameAnimationActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.frame_animations_layout)
        setupButton()
    }

    private fun setupButton() {
        val b = findViewById<Button>(R.id.startFAButtonId)
        b.setOnClickListener { parentButtonClicked() }
    }

    private fun parentButtonClicked() {
        animate()
    }

    private fun animate() {
        val imgView = findViewById<ImageView>(R.id.imageView)
        imgView.visibility = ImageView.VISIBLE
        imgView.setBackgroundResource(R.drawable.frame_animation)

        val frameAnimation = imgView.background as AnimationDrawable
        if (frameAnimation.isRunning) {
            frameAnimation.stop()
        } else {
            frameAnimation.stop()
            frameAnimation.start()
        }
    }

}